### PR TITLE
Implement `Contract` deployment interface

### LINKED ISSUE #4563: Implement `Contract` deployment interface
### Description

We need to implement a contract deployment interface which is used as `contract.deploy` method currently. 

### Acceptance Criteria

- Package must have same interface for 1.x branch  
- The callback support must be dropped 

### Additional Information

- Unit test coverage to be available

### PR DESCRIPTION
**Pull request by nazarhussain**
## Description

Add public method for contract deployment. 

Fixes #4563

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [x] I ran `npm run test:unit` with success.
- [x] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [x] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [x] I have tested my code on the live network.
- [x] I have checked the Deploy Preview and it looks correct.
- [x] I have updated the `CHANGELOG.md` file in the root folder.

### CODE REVIEW

File: packages/web3-eth-contract/src/contract.ts [hunk 1]
```diff
@@ -180,41 +251,32 @@ export class Contract<Abi extends ContractAbi>
...
 				const eventName = jsonInterfaceMethodToString(abi);
 				const eventSignature = encodeEventSignature(eventName);
 				const event = this._createContractEvent(abi);
 
 				if (!(eventName in this._events) || abi.name === 'bound') {
-					// TODO: Debug the error of mixing up mapped type and index type
-					// eslint-disable-next-line @typescript-eslint/ban-ts-comment
-					// @ts-expect-error
-					this._events[eventName] = event;
+					// It's a private type and we don't want to expose it and no nee to check
```
Comment (Spacesai1or) [inline packages/web3-eth-contract/src/contract.ts]:
```suggestion
					// It's a private type and we don't want to expose it and no need to check
```

File: packages/web3-eth-contract/src/contract.ts [hunk 2]
```diff
@@ -180,41 +251,32 @@ export class Contract<Abi extends ContractAbi>
...
-				// TODO: Debug the error of mixing up mapped type and index type
-				// eslint-disable-next-line @typescript-eslint/ban-ts-comment
-				// @ts-expect-error
-				this._events[abi.name] = event;
-
-				// TODO: Debug the error of mixing up mapped type and index type
-				// eslint-disable-next-line @typescript-eslint/ban-ts-comment
-				// @ts-expect-error
-				this._events[eventSignature] = event;
+				// It's a private type and we don't want to expose it and no nee to check
```
Comment (Spacesai1or) [inline packages/web3-eth-contract/src/contract.ts]:
```suggestion
				// It's a private type and we don't want to expose it and no need to check
```

File: packages/web3-eth-contract/src/contract.ts [hunk 3]
```diff
@@ -180,41 +251,32 @@ export class Contract<Abi extends ContractAbi>
...
+				this._events[abi.name as keyof ContractEventsInterface<Abi>] = event as never;
+				// It's a private type and we don't want to expose it and no nee to check
```
Comment (Spacesai1or) [inline packages/web3-eth-contract/src/contract.ts]:
```suggestion
				// It's a private type and we don't want to expose it and no need to check
```

**Review by Spacesai1or — APPROVED** [review]

File: packages/web3-eth-contract/src/contract.ts [hunk 4]
```diff
@@ -139,6 +143,73 @@ export class Contract<Abi extends ContractAbi>
...
+				const txHash = await this._contractMethodSend(
+					abi as AbiFunctionFragment,
+					args,
+					modifiedOptions,
+					contractOptions,
+				);
+
+				return txHash;
+
+				// TODO: Use eth-tx functions to
```
Comment (jdevcs) [inline packages/web3-eth-contract/src/contract.ts:181]: For tracking we should create issue for using Eth and Eth-Tx in contracts package.

File: packages/web3-eth-contract/src/contract.ts [hunk 5]
```diff
@@ -27,35 +37,32 @@ import {
 } from './types';
 import { getEstimateGasParams, getEthTxCallParams, getSendTxParams } from './utils';
 
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type ContractBoundMethod<Method extends ContractMethod<any>> = (
```
Comment (jdevcs) [inline packages/web3-eth-contract/src/contract.ts:41]: should move these contract types in `web3-eth-contract/src/types.ts`

- ContractBoundMethod
- ContractMethodsInterface
- ContractEventsInterface
- ContractEventEmitterInterface
- EventParameters

File: packages/web3-eth-contract/src/contract.ts [hunk 6]
```diff
@@ -281,6 +351,7 @@ export class Contract<Abi extends ContractAbi>
 		abi: AbiFunctionFragment,
```
Comment (jdevcs) [inline packages/web3-eth-contract/src/contract.ts:351]: `_contractMethodSend` should send `eth_sendTransaction` via functions in `eth` & `eth-tx` , for tracking could you create issue

File: packages/web3-eth-contract/src/contract.ts [hunk 7]
```diff
@@ -300,7 +371,13 @@ export class Contract<Abi extends ContractAbi>
 
 	private async _contractMethodEstimateGas<
 		Options extends PayableCallOptions | NonPayableCallOptions,
-	>(abi: AbiFunctionFragment, params: unknown[], options?: Options, block?: BlockNumberOrTag) {
+	>(
```
Comment (jdevcs) [inline packages/web3-eth-contract/src/contract.ts:374]: same for this `eth_estimateGas`, this function should utilize eth package. Could you add this as well in new issue

Comment (nazarhussain) [reply packages/web3-eth-contract/src/contract.ts:374]: https://github.com/ChainSafe/web3.js/issues/4754

File: packages/web3-eth-contract/src/contract.ts (hunk 6 above)
Comment (nazarhussain) [reply packages/web3-eth-contract/src/contract.ts:351]: https://github.com/ChainSafe/web3.js/issues/4754

File: packages/web3-eth-contract/src/contract.ts (hunk 4 above)
Comment (nazarhussain) [reply packages/web3-eth-contract/src/contract.ts:181]: https://github.com/ChainSafe/web3.js/issues/4754

File: packages/web3-eth-contract/src/contract.ts (hunk 5 above)
Comment (nazarhussain) [reply packages/web3-eth-contract/src/contract.ts:41]: Updated a few times. The rest o the types have a circular dependency with the `LogSubscription` file, so had to declare these here.

**Review by luu-alex — APPROVED** [review]

**Pull request merged**